### PR TITLE
KZOO-149: Add sftp support

### PR DESCRIPTION
**Pull request by holy-batman**
This adds support for ssh servers to be uses a media saving locations, the ssh servers are automatically added to known_host file.

This module has been tested with a ssh server, using username and password for authentication to save and retrieve media.

Url example: `sftp://user:pass@host_url:port/foder_location`

### CODE REVIEW

**Review by lazedo — CHANGES_REQUESTED** [review]
add sftp in a new module, don't touch the ftp module.
make sure you test this with data plans, ensure that you can save & read an attachment

**Comment by holy-batman** [comment]
@lazedo changes have been made as requested.

**Comment by holy-batman** [comment]
requires https://github.com/2600hz/kazoo-media-mgr/pull/10

File: kazoo_attachments/src/kz_att_sftp.erl [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+fetch_attachment(Url) ->
+    case uri_string:parse(kz_term:to_list(Url)) of
+        #{path := FullPath} = URI ->
+            get_sftp_request(sftp_connect(URI), FullPath);
+        _ -> {'error', <<"error parsing url: ", Url/binary>>}
+    end.
+
+get_sftp_request({Host, Port, Options}, FullPath) ->
+    try ssh_sftp:start_channel(Host, Port, Options) of
+        {'ok', Pid, _} -> handle_fetch(Pid, ssh_sftp:read_file(Pid, FullPath, Options));
```
Comment (jamesaimonetti) [inline kazoo_attachments/src/kz_att_sftp.erl]: `Options` should be a `Timeout` according to the spec:

```erlang
-spec read_file(ChannelPid, File, Timeout) -> {ok, Data} | Error when
      ChannelPid :: pid(),
      File :: string(),
      Data :: binary(),
      Timeout :: timeout(),
      Error :: {error, reason()}.
read_file(Pid, Name, FileOpTimeout) ->
```

**Review by jamesaimonetti — CHANGES_REQUESTED** [review]

File: kazoo_attachments/src/kz_att_sftp.erl [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+            {'error', Err}
+    end.
+
+-type sftp_connect_info() :: {string(), pos_integer(),
+                              [{'timeout', pos_integer()} | {user, string()} | {password, string()} | {'silently_accept_hosts', boolean()}]}.
+
+-spec sftp_connect(map()) -> sftp_connect_info().
+sftp_connect(#{host := Host, port := Port, userinfo := UserPass}) ->
+    [User, Pass] = string:tokens(UserPass, ":"),
+    {Host, Port, [{'timeout', 5000}, {'user', User}, {'password', Pass}, {'silently_accept_hosts', 'true'}]};
```
Comment (jamesaimonetti) [inline kazoo_attachments/src/kz_att_sftp.erl]: Should be

```erlang
{Host, Port
 ,[{'timeout', 5000}
   ,{'user', User}
   ,{'password', Pass}
   ,{'silently_accept_hosts', 'true'}
  ]
 }
```

File: kazoo_attachments/src/kz_att_sftp.erl [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+sftp_connect(#{host := Host, userinfo := UserPass}) ->
+    [User, Pass] = string:tokens(UserPass, ":"),
+    {Host, 22, [{'timeout', 5000}, {'user', User}, {'password', Pass}, {'silently_accept_hosts', 'true'}]}.
```
Comment (jamesaimonetti) [inline kazoo_attachments/src/kz_att_sftp.erl]: Let's also look for timeout in the storage plan options and, if not present, default to 10s (please use `10 * ?MILLISECONDS_IN_SECOND` so we don't have magic numbers floating around).

File: kazoo_attachments/src/kz_att_sftp.erl (hunk 2 above)
Comment (holy-batman) [reply kazoo_attachments/src/kz_att_sftp.erl]: done :)

**Pull request merged**